### ISSUE TITLE
Unstable dask._Frame.map_partitions behavior

### ISSUE THREAD

**Issue by sscondie** [blaze/dask#888]
Consider the following script operating on 15 csv files each about 16MB in size (each csv is an array of floats that is 10 columns and 100,000 rows).  

```python
import dask.dataframe as dd
df = dd.read_csv("test_data_small/*.csv")
print(df.npartitions)
print(df.mean().compute())
print(df.map_partitions(len,columns=df.columns).compute())
```

Running this script on my machine (OS X (10.11.1), Python 2.7.11 [Anaconda 2.3.0 (x86_64)], pandas 0.17.1) produces (in approximately the same frequencies and in random order) one of the following results (for each result I give first a description and then stdout)

**1. npartitions completes, mean completes, map_partitions seg faults** 

```text
15
0                -0.000135
1                -0.000011
10               -0.000939
2                 0.000771
3                 0.000309
4                 0.000573
5                 0.000635
6                 0.000770
7                 0.000713
8                 0.001237
9                -0.000440
Unnamed: 0    49999.500000
dtype: float64
Segmentation fault: 11
```
**2. npartitions completes, Fatal Python Error at mean**

```text
15
Fatal Python error: GC object already tracked
Abort trap: 6
```

**3. npartitions completes, mean completes, Fatal Python Error at map_partitions**
 
```text
15
0                -0.000135
1                -0.000011
10               -0.000939
2                 0.000771
3                 0.000309
4                 0.000573
5                 0.000635
6                 0.000770
7                 0.000713
8                 0.001237
9                -0.000440
Unnamed: 0    49999.500000
dtype: float64
Fatal Python error: GC object already tracked
Abort trap: 6
```

**4.  npartitions completes, seg fault at mean**

```text
15
Segmentation fault: 11
```

**5.  Everything completes**

```text
15
0                -0.000135
1                -0.000011
10               -0.000939
2                 0.000771
3                 0.000309
4                 0.000573
5                 0.000635
6                 0.000770
7                 0.000713
8                 0.001237
9                -0.000440
Unnamed: 0    49999.500000
dtype: float64
(100001, 100001, 100001, 100001, 100001, 100001, 100001, 100001, 100001, 100001, 100001, 100001, 100001, 100001, 100001)
```

I reverted pandas back to 0.17.0 based on a suggestion from Matt [here] http://stackoverflow.com/questions/34128540/why-is-running-a-compute-in-dask-causing-fatal-python-error-gc-object-alrea  but that hasn't noticeably changed the results.

I'm happy to debug this in dask, but there is at least some chance that this could be a pandas problem.  Has anyone else seen this kind of behavior?  If so, were you able to find an answer?

**Comment by sscondie** [comment]
Weird numbering, sorry.  Markdown has had its way with me somehow.

**Comment by sscondie** [comment]
Update:  The problem seems to be related to the threaded scheduler.  When I explicitly set the scheduler using get = dask.multiprocessing.get then all functions complete.  When I set it back to get=dask.threaded.get then I get the problems above.

**Comment by jcrist** [comment]
I haven't seen segfaults before with pandas 0.17.0 (as you noted, some issues have come up with pandas 0.17.1), and am unable to reproduce with files I created on my machine. It may be an issue with the data you have. Either way, the segfaulting is caused by pandas not being threadsafe, and will need to be fixed on their end.

Perhaps try loading in data into a pandas dataframes in serial, then see if you still have issues with segfaulting. I suspect the problem is in `read_csv`.

```python
import pandas as pd
import dask.dataframe as dd
from glob import glob
dfs = [pd.read_csv(f) for f in glob('test_data_small/*.csv')]
df = dd.concat([dd.from_pandas(i, npartitions=1) for i in dfs])
print(df.npartitions)
print(df.mean().compute())
print(df.map_partitions(len,columns=df.columns).compute())
```

P.S. I cleaned up the markdown for you - wrap code samples with 3 backticks to improve legibility (```). Numbering lists also has its quirks.

**Comment by mrocklin** [comment]
Pandas 0.17.1 introduced some unsafe code into `read_csv`.  Relevant Pandas issue here https://github.com/pydata/pandas/issues/11786

A simple fix was introduced in https://github.com/pydata/pandas/pull/11790 but has yet to be merged.  It may be fruitful to have others ping that issue.

**Comment by sscondie** [comment]
Thanks @jcrist.  It doesn't appear to be a pandas problem in the following sense:  This code:

```text
file_list = glob.glob( "test_data_small/*.csv")
for ff in file_list:
    with open(ff,'r') as fh:
        df = pd.read_csv(fh)
        print(df.mean())
```

runs without a hitch.  Could this possibly be related to issue [878](https://github.com/blaze/dask/issues/878) about how dask misinterprets pandas-generated column names?  I generated the data from pandas.DataFrame.to_csv().

**Comment by jcrist** [comment]
What I meant was (and also as @mrocklin says above), that `pd.read_csv` isn't threadsafe in 0.17.1, and so reading csvs in parallel is (probably) causing the issue. The non-threadsafe nature of `read_csv` is what's a pandas problem.

I haven't seen this behavior in pandas 0.17.0 though, which is why I'm confused, as you said you tried this with 0.17.0 as well and still saw the same issues.

**Comment by sscondie** [comment]
Yes, this is all the under 0.17.0 as mentioned. 






On Tue, Dec 22, 2015 at 8:04 PM -0800, "Jim Crist" <notifications@github.com> wrote:












What I meant was (and also as @mrocklin says above), that pd.read_csv isn't threadsafe in 0.17.1, and so reading csvs in parallel is (probably) causing the issue. The non-threadsafe nature of read_csv is what's a pandas problem.



I haven't seen this behavior in pandas 0.17.0 though, which is why I'm confused, as you said you tried this with 0.17.0 as well and still saw the same issues.

**Comment by mrocklin** [comment]
Ah, my apologies for missing the note above.  I was going quickly through my e-mails.  Two things to check.

1.  If you are comfortable running from a development branch then try the code in the PR pointed to in my last comment and see if the problem persists.
2.  Verify that everything works fine if you use the non-threaded scheduler with `...compute(get=dask.get)`

It is unlikely that dask alone is responsible for a segfault; it's pure Python.  Problems like this usually arise when we incorrectly assume that a C-calling library is threadsafe.  Pandas has been really good recently about releasing the GIL, allowing multi-threaded code (like dask) to accelerate pandas computations.  It's not surprising that this development should fail from time to time.  Experiments like yours help both development teams to identify failures.

**Comment by sscondie** [comment]
It completes with ...compute(get=dask.get).  I'll check back after I've tried the referenced pandas PR.

**Comment by mrocklin** [comment]
What is the status of this?

**Comment by mrocklin** [comment]
Relevant pandas issues were merged.  I'm going to assume that this is complete.  Please reopen if the problem persists.

**Comment by QCaudron** [comment]
I'm not sure this is the same issue, but I get the same error followed by a segfault when I try returning the length of a `dask.dataframe` :

    Fatal Python error: GC object already tracked
    /Users/qcaudron/anaconda/bin/python.app: line 3: 40449 Abort trap: 6           /Users/qcaudron/anaconda/python.app/Contents/MacOS/python "$@"

I'm running dask 0.8.0 and pandas 0.17.1 through Anaconda.

**Comment by mrocklin** [comment]
Can you try updating to the Pandas release candidate, version 0.18.0?  A lot of pandas multi-threading issues were cleaned up in the last release.

Using conda

    conda install -c pandas pandas=v0.18.0rc1

Using pip

    pip install git+https://github.com/pydata/pandas.git --upgrade

**Comment by QCaudron** [comment]
Thanks for the super fast answer - I was already updating to RC 0.18 after reading up pydata/pandas#11790. `len(df)` works fine now !
